### PR TITLE
refactor(hfx)!: remove `fastify`

### PR DESCRIPTION
**Pull request by kyranet**
While `fastify` is a very high quality module with little to no overhead compared to raw HTTP, it requires some hacks that would otherwise not be needed in raw HTTP due to the high level of abstraction.

And because it's not recommended to have plugins or anything that might extend the framework's HTTP listener, it was not necessary to have such a nice wrapper.

On a side note, this has shaved 12 MB of memory usage from Teryl. 👀

### CODE REVIEW

File: packages/http-framework/src/lib/Client.ts [hunk 1]
```diff
@@ -52,29 +54,68 @@ export class Client extends EventEmitter {
...
 
-		this.server = Fastify(serverOptions);
-		this.server.addHook('onRoute', (options) => assignRawData(options));
-		this.server.addHook('preHandler', (request, reply) => handleSecurityHook(request, reply, key));
-		this.server.post(postPath ?? process.env.HTTP_POST_PATH ?? '/', this.handleHttpMessage.bind(this));
+		this.server = createServer(serverOptions ?? {});
+		this.server.on('request', (request, response) => this.handleRawHttpMessage(request, response, path, key));
 
-		await this.server.listen({ ...listenOptions, port, host: address });
+		return new Promise<void>((resolve) => this.server.listen({ ...listenOptions, port, host: address }, resolve));
```
Comment (vladfrangu) [inline packages/http-framework/src/lib/Client.ts:62]: Doesn't this silence errors that happen when listening?

File: packages/http-framework/src/lib/Client.ts [hunk 2]
```diff
@@ -52,29 +54,68 @@ export class Client extends EventEmitter {
...
 
-	protected async handleHttpMessage(request: FastifyRequest, reply: FastifyReply): Promise<FastifyReply> {
-		const interaction = request.body as APIInteraction;
-		if (interaction.type === InteractionType.Ping) return reply.send({ type: InteractionResponseType.Pong });
+	protected async handleRawHttpMessage(request: IncomingMessage, response: ServerResponse, path: string, key: Key) {
+		response.setHeader('Content-Type', 'application/json');
+
+		if (request.url !== path) {
+			response.statusCode = HttpCodes.NotFound;
+			return response.end('{"message":"Unsupported method"}');
```
Comment (vladfrangu) [inline packages/http-framework/src/lib/Client.ts:70]: At least have a nicer 404 message than the same one as 405

File: packages/http-framework/src/lib/Client.ts [hunk 3]
```diff
@@ -52,29 +54,68 @@ export class Client extends EventEmitter {
...
+			return response.end('{"message":"The signature is incorrect"}');
+		}
+
+		return this.handleHttpMessage(JSON.parse(body) as APIInteraction, response);
+	}
+
+	protected async handleHttpMessage(interaction: APIInteraction, response: ServerResponse): Promise<ServerResponse> {
+		if (interaction.type === InteractionType.Ping) {
+			response.statusCode = HttpCodes.OK;
+			return response.end(`{"type":${InteractionResponseType.Pong}}`);
```
Comment (vladfrangu) [inline packages/http-framework/src/lib/Client.ts:104]: Please don't use json strings, but JSON.stringified objects. Last thing you need is debugging a missing quote

File: packages/http-framework/src/lib/interactions/structures/Interaction.ts [hunk 4]
```diff
@@ -263,9 +263,12 @@ export class Interaction<I extends BaseInteractionType = BaseInteractionType> {
...
 
-	protected async _sendReply(data: NonNullObject) {
-		if (this[Reply].sent) throw new Error('Cannot send response, the request has already been replied.');
-		await this[Reply].status(HttpCodes.OK).type('application/json').send(data);
+	protected _sendReply(data: NonNullObject) {
+		const response = this[Response];
+		if (response.closed) throw new Error('Cannot send response, the request has already been replied.');
+
+		response.statusCode = HttpCodes.OK;
+		return new Promise<void>((resolve) => response.end(JSON.stringify(data), resolve));
```
Comment (vladfrangu) [inline packages/http-framework/src/lib/interactions/structures/Interaction.ts:271]: What's the resolve here for? Also does it silence errors?

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: packages/http-framework/src/lib/Client.ts (hunk 1 above)
Comment (kyranet) [reply packages/http-framework/src/lib/Client.ts:62]: Errors are thrown synchronously, `listen(option, cb)` does not pass any parameters (not even an error) to the callback.

File: packages/http-framework/src/lib/interactions/structures/Interaction.ts (hunk 4 above)
Comment (kyranet) [reply packages/http-framework/src/lib/interactions/structures/Interaction.ts:271]: It's called when the data has been fully sent, see [docs](https://nodejs.org/dist/latest-v18.x/docs/api/stream.html#writableendchunk-encoding-callback).

**Review by vladfrangu — APPROVED** [review]

**Review by favna — APPROVED** [review]

**Pull request merged**